luigi/contrib/htcondor.py

```diff
@@ -0,0 +1,370 @@
+        # add job_file to input files
+        # enable transfer of input files
+        # create script that
+        #  - calls the user pre-script
+        #  - python <runner> <pickle>
+        #  - calls the user post-script
+        #
+        _copy_script_to_tmp_dir(self.pre_script, self.tmp_dir, 'pre.sh')
+        _copy_script_to_tmp_dir(self.post_script, self.tmp_dir, 'post.sh')
+        runner_path = sge_runner.__file__
+        _copy_script_to_tmp_dir(runner_path, self.tmp_dir, 'runner.py')
+
+        tarball_param = ""
+        if self.no_tarball:
+            tarball_param = "--no-tarball"
+
+        run_sh = os.path.join(self.tmp_dir, 'run.sh')
+        with open(run_sh, 'w') as f:
+            f.write(RUN_SCRIPT.format(tarball_param=tarball_param))
+        st = os.stat(run_sh)
+        os.chmod(run_sh, st.st_mode | stat.S_IEXEC)
+
+        self.outfile = os.path.join(self.tmp_dir, 'job.out')
+        self.errfile = os.path.join(self.tmp_dir, 'job.err')
+        job_params = {
+            'Cmd': run_sh,
+            'TransferIn': ','.join(glob.glob(self.tmp_dir + '/*')),
+            'Out': self.outfile,
+            'Err': self.errfile,
+            'UserLog': os.path.join(self.tmp_dir, 'job.log'),
+            'request_cpus': self.n_cpu,
+            'MemoryUsage': self.memory,
+        }
+        job_params.update(self.additional_hooks)
+        # build job description (mostly for debugging)
+        job_desc = _build_job_description(job_params)
+        logger.debug('Submitting htcondor job description: \n' + str(job_desc))
+        from htcondor import Schedd
+        # connect to HTCondor scheduler
+        schedd = Schedd()
+        # Submit the job and grab job ID.
+        self.job_id = schedd.submit(job_desc)
+        logger.debug(
+            "Submitted job to condor_submit with ID:\n{0}".format(self.job_id))
+
+        self._track_job()
+
+        # Now delete the temporaries, if they're there.
+        # unless transfer_output files is on.
+        if os.path.exists(self.tmp_dir) and not self.dont_remove_tmp_dir:
+            logger.info('Removing temporary directory %s' % self.tmp_dir)
+            subprocess.call(["rm", "-rf", self.tmp_dir])
+
+    def _track_job(self):
+        """ Known job statuses
+        0   Unexpanded     U
+        1   Idle           I
+        2   Running        R
+        3   Removed        X
+        4   Completed      C
+        5   Held           H
+        6   Submission_err E
+        < transferring input
+        > transferring output
+        """
+        from htcondor import Schedd
+        known_statuses = [
+            'Unexpanded', 'Idle', 'Running', 'Removed', 'Completed', 'Held',
+            'Error', '<', '>'
+        ]
+        schedd = Schedd()
+        contraint = 'JobId =?= {0}'.format(self.job_id)
+        while True:
+            # Sleep for a little bit
+            time.sleep(self.poll_time)
+            job_ad = schedd.query(contraint)
```

Comment:
Also, use `xquery` instead - it'll return an iterator that emits the resulting ads..